File: src/api/events.ts
import { listen, type UnlistenFn } from '@tauri-apps/api/event'
import type { SetupDriveFile } from './setup'
import type { SyncStatus } from './sync'
import type { EntryMeta } from './types'

/**
 * The event catalog. The backend `emit`s these; the frontend `listen`s. Used for
 * background work (sync, auto-lock, a long import) that is not a direct
 * request/response command.
 *
 * `setup:drive:pending` → `setup:drive:probed` | `setup:drive:error` is a
 * consent flow out with the browser against an account there is no vault behind
 * yet. The backend owns all three: it is what opens the browser and what hears
 * back from it, so the frontend mirrors these rather than guessing from a click
 * — the command itself resolves the moment the browser is on screen. An empty
 * `files` there is an account with no Rowel data in it, a fact rather than a
 * failure, so it is not an error event.
 */
export interface EventPayloads {
  'vault:locked': void
  /**
   * A sync pulled entries this device did not have. Carries the whole refreshed
   * list rather than a "reload" ping so the store updates in one render, and
   * because the backend has already paid for the query.
   */
  'vault:merged': { entries: EntryMeta[] }
  /**
   * The whole of sync, every time anything about it changes: a consent flow
   * opening or closing, a run starting or ending. The frontend stores it as-is
   * — there is no sequence of events to reconstruct state from.
   */
  'sync:status': SyncStatus
  'import:progress': { done: number; total: number }
  'setup:drive:pending': void
  /** Every vault the account holds, newest first; empty is "nothing to restore". */
  'setup:drive:probed': { files: SetupDriveFile[] }
  'setup:drive:error': { error: string }
  /**
   * After every sync run: the account's vaults that no workspace on this
   * device holds, newest first, in the probe's shape because Settings ›
   * Workspaces offers them through the same picker. Empty means every vault in
   * the account is already here.
   */
  'workspaces:remote': { files: SetupDriveFile[] }
  /**
   * A vault from the account was added as a workspace without being asked
   * for: the password that just opened a vault opened it too. Nothing has
   * switched; the list of workspaces has grown by `name`.
   */
  'workspaces:added': { name: string }
  /**
   * A sync pulled a newer name for the open workspace — it was renamed on
   * another device. No payload: names live on `app_status`, so the frontend
   * re-reads that and the header and Workspaces list redraw from it.
   */
  'workspaces:renamed': void
  /**
   * The OS opened a backup with the app — a double-clicked `.rowel` or
   * `.swftx`. Rust also parks the path for a shell that was not yet listening
   * (`takeOpenedFile` in `api/app`), so a launch by double-click and an open
   * while running both end up in `store/app`'s `fileOpened`.
   */
  'file:opened': { path: string }
  /**
   * A browser extension asks to be let in. `key` is its identification public
   * key, base64, for the consent dialog to show a fingerprint of; the answer
   * goes back through `browser_respond`. Desktop only.
   */
  'browser:associate': { key: string }
  /**
   * An extension was let into the open vault by the consent dialog. Settings ›
   * Browser extension re-reads its status on it; no payload, since that status
   * is the one answer its list is drawn from. Desktop only.
   */
  'browser:clients': void
}

export type EventName = keyof EventPayloads

type Camel<S extends string> = S extends `${infer Head}:${infer Tail}`
  ? `${Head}${Capitalize<Camel<Tail>>}`
  : S

/**
 * Call-site names for the events. Derived from `EventPayloads`, so adding an
 * event without a key here — or keeping a key whose event is gone — fails the
 * build rather than drifting.
 */
export const EVENTS: { [K in EventName as Camel<K>]: K } = {
  vaultLocked: 'vault:locked',
  vaultMerged: 'vault:merged',
  syncStatus: 'sync:status',
  importProgress: 'import:progress',
  setupDrivePending: 'setup:drive:pending',
  setupDriveProbed: 'setup:drive:probed',
  setupDriveError: 'setup:drive:error',
  workspacesRemote: 'workspaces:remote',
  workspacesAdded: 'workspaces:added',
  workspacesRenamed: 'workspaces:renamed',
  fileOpened: 'file:opened',
  browserAssociate: 'browser:associate',
  browserClients: 'browser:clients'
}

// Typed wrapper over Tauri's `listen`.
export const on = <E extends EventName>(
  event: E,
  handler: (payload: EventPayloads[E]) => void
): Promise<UnlistenFn> => listen<EventPayloads[E]>(event, e => handler(e.payload))

File: src-tauri/src/events.rs
//! Every event the backend pushes at the frontend, and the only way to send one.
//!
//! Names and payload shapes live here rather than at the call sites so the
//! catalogue the frontend subscribes to can be read in one place, and so no
//! caller can invent a third spelling of `{ error }`.

use serde::Serialize;
use tauri::{AppHandle, Emitter};

use crate::models::EntryMetaDto;
use crate::state::SyncStatus;
use crate::sync::setup::PackInfo;

pub const VAULT_LOCKED: &str = "vault:locked";
pub const VAULT_MERGED: &str = "vault:merged";
pub const SYNC_STATUS: &str = "sync:status";
pub const IMPORT_PROGRESS: &str = "import:progress";
pub const SETUP_DRIVE_PENDING: &str = "setup:drive:pending";
pub const SETUP_DRIVE_PROBED: &str = "setup:drive:probed";
pub const SETUP_DRIVE_ERROR: &str = "setup:drive:error";
pub const WORKSPACES_REMOTE: &str = "workspaces:remote";
pub const WORKSPACES_ADDED: &str = "workspaces:added";
pub const WORKSPACES_RENAMED: &str = "workspaces:renamed";
// Desktop only, like the file associations that produce it (`crate::opened`).
#[cfg(desktop)]
pub const FILE_OPENED: &str = "file:opened";
// Desktop only, like the extension host that raises it (`crate::browser`).
#[cfg(desktop)]
pub const BROWSER_ASSOCIATE: &str = "browser:associate";

#[derive(Serialize, Clone)]
struct Entries {
    entries: Vec<EntryMetaDto>,
}

#[cfg(desktop)]
#[derive(Serialize, Clone)]
struct Opened<'a> {
    path: &'a str,
}

#[derive(Serialize, Clone)]
struct ErrorText<'a> {
    error: &'a str,
}

#[derive(Serialize, Clone)]
struct Progress {
    done: usize,
    total: usize,
}

/// Every vault the account holds, so the first run can offer a choice rather
/// than one picked for it. An empty list is an answer — the account holds no
/// vault — not an absence.
#[derive(Serialize, Clone)]
struct Probed {
    files: Vec<PackInfo>,
}

pub fn vault_locked(app: &AppHandle) {
    let _ = app.emit(VAULT_LOCKED, ());
}

pub fn vault_merged(app: &AppHandle, entries: Vec<EntryMetaDto>) {
    let _ = app.emit(VAULT_MERGED, Entries { entries });
}

/// The whole of sync, every time anything about it changes. One snapshot
/// rather than one event per transition: the frontend stores it as-is, and
/// neither side has to agree on an order of events.
pub fn sync_status(app: &AppHandle, status: SyncStatus) {
    let _ = app.emit(SYNC_STATUS, status);
}

pub fn import_progress(app: &AppHandle, done: usize, total: usize) {
    let _ = app.emit(IMPORT_PROGRESS, Progress { done, total });
}

pub fn setup_drive_pending(app: &AppHandle) {
    let _ = app.emit(SETUP_DRIVE_PENDING, ());
}

pub fn setup_drive_probed(app: &AppHandle, files: Vec<PackInfo>) {
    let _ = app.emit(SETUP_DRIVE_PROBED, Probed { files });
}

pub fn setup_drive_error(app: &AppHandle, error: &str) {
    let _ = app.emit(SETUP_DRIVE_ERROR, ErrorText { error });
}

/// The account's vaults that no workspace on this device holds, after every
/// sync run — the same shape as the probe's answer, because Settings ›
/// Workspaces offers them through the same picker. Empty means every vault in
/// the account is here.
pub fn remote_vaults(app: &AppHandle, files: Vec<PackInfo>) {
    let _ = app.emit(WORKSPACES_REMOTE, Probed { files });
}

#[derive(Serialize, Clone)]
struct Added<'a> {
    name: &'a str,
}

/// A vault from the account was added as a workspace without being asked for
/// — the password that just opened a vault opened it too (`commands::autojoin`).
/// The frontend re-probes for the new list and says so; nothing switches.
pub fn workspace_added(app: &AppHandle, name: &str) {
    let _ = app.emit(WORKSPACES_ADDED, Added { name });
}

/// A sync pulled a newer name for the open workspace. No payload: the names
/// live on `app_status`, which the frontend re-reads, so the header and the
/// Workspaces list redraw from the one answer rather than from two.
pub fn workspace_renamed(app: &AppHandle) {
    let _ = app.emit(WORKSPACES_RENAMED, ());
}

/// The OS asked the app to open a backup (see `crate::opened`). Also parked for
/// a webview that is not listening yet; this is the live half.
#[cfg(desktop)]
pub fn file_opened(app: &AppHandle, path: &str) {
    let _ = app.emit(FILE_OPENED, Opened { path });
}

#[cfg(desktop)]
#[derive(Serialize, Clone)]
struct Associate<'a> {
    id: &'a str,
    key: &'a str,
}

/// A browser extension asks to be let in (`browser::consent`). `key` is its
/// identification public key, for the dialog to show a fingerprint of; the
/// answer comes back through `commands::browser::browser_respond`, naming
/// `id` — this ask's alone, so a dialog left up past it cannot answer the
/// next ask the same extension makes.
#[cfg(desktop)]
pub fn browser_associate(app: &AppHandle, id: &str, key: &str) {
    let _ = app.emit(BROWSER_ASSOCIATE, Associate { id, key });
}
